### PR TITLE
Unit tests for transformer ranking

### PR DESCRIPTION
**Pull request by Jerry-Jinfeng-Guo**

### CODE REVIEW

File: power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp [hunk 1]
```diff
@@ -120,9 +120,11 @@ inline auto rank_transformers(WeightedTrafoList const& w_trafo_list) -> RankedTr
               [](const WeightedTrafo& a, const WeightedTrafo& b) { return a.second < b.second; });
 
     RankedTransformerGroups groups;
+    std::vector<Idx> weights;
     for (const auto& trafo : sorted_trafos) {
-        if (groups.empty() || groups.back().back().pos != trafo.second) {
+        if (groups.empty() || weights.back() != trafo.second) {
             groups.push_back(std::vector<Idx2D>{trafo.first});
+            weights.push_back(trafo.second);
```
Comment (Jerry-Jinfeng-Guo) [inline power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp:127]: Crime scene of `Idx2D` biting

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+#include <power_grid_model/optimizer/tap_position_optimizer.hpp>
+
+#include <boost/graph/graph_utility.hpp>
+
+#include <doctest/doctest.h>
+
+namespace pgm_tap = power_grid_model::optimizer::tap_position_optimizer;
+
+namespace power_grid_model {
+namespace {} // namespace
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:14]: either fill or remove

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    pgm_tap::TransformerGraph g;
+    //{boost::edges_are_unsorted_multi_pass, edge_array.cbegin(), edge_array.cend(), 3};
+
+    SUBCASE("Process edge weights") { auto const res = get_edge_weights(g); }
+
+    SUBCASE("Dijkstra shortest path") {
+        std::vector<pgm_tap::EdgeWeight> edge_weight(3, pgm_tap::infty);
+        std::vector<Idx2D> edge_pos(3);
+        // process_edges_dijkstra(0, edge_weight, edge_pos, g);
+    }
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:37]: i don't necessarily think that there's there's a need to determine shortest path itself. just a check that the input and output of the ranking is correct. all the rest is implementation detail (if you prefer, you can put it in `namespace detail`)

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace power_grid_model {
+
+TEST_CASE("Test Transformer ranking") {
+    // ToDo: The grid from OntNote page
+
+    // Subcases
+    SUBCASE("Building the graph") {
+        // ToDo: graph creation
+    }
```
Comment (Jerry-Jinfeng-Guo) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:27]: These two parts will be added soon

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+namespace pgm_tap = power_grid_model::optimizer::tap_position_optimizer;
+
+namespace power_grid_model {
+
+TEST_CASE("Test Transformer ranking") {
+    // ToDo: The grid from OntNote page
+
+    // Subcases
+    SUBCASE("Building the graph") {
+        // ToDo: graph creation
```
Comment (mgovers) [inline tests/cpp_unit_tests/test_tap_position_optimizer.cpp:18]: pro-tip: if you all-caps `TODO` or `FIXME`, then most editors will highlight it. In addition, SonarCloud will also be able to understand it

**Review by mgovers — APPROVED** [review]
LGTM. I guess that you will add the rest in a separate PR?

File: tests/cpp_unit_tests/test_tap_position_optimizer.cpp (hunk 5 above)
Comment (Jerry-Jinfeng-Guo) [reply tests/cpp_unit_tests/test_tap_position_optimizer.cpp:18]: Ok, that answers one of my lingering question...

**Comment by Jerry-Jinfeng-Guo** [comment]
> LGTM. I guess that you will add the rest in a separate PR?

Yup, I think so. As soon as it could be added.

**Pull request merged**